### PR TITLE
WildFly 10 example

### PR DESCRIPTION
**Pull request by icetrain**

### CODE REVIEW

File: example_configs/wildfly-10.yaml [hunk 1]
```diff
@@ -0,0 +1,35 @@
+---
+lowercaseOutputName: true
+lowercaseOutputLabelNames: true
+whitelistObjectNames:
```
Comment (brian-brazil) [inline example_configs/wildfly-10.yaml]: Given this is a framework, you shouldn't assume only the wildfly objects are of potential interest

File: example_configs/wildfly-10.yaml [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+  - "jboss.as:subsystem=messaging-activemq,server=*,jms-topic=*"
+  - "jboss.as:subsystem=transactions"
+  - "jboss.as:subsystem=undertow,server=*,http-listener=*"
+  
+rules:
+  - pattern: "^jboss.as<subsystem=messaging-activemq, server=.+, jms-(?:queue|topic)=(.+)><>(.+):"
+    attrNameSnakeCase: true
+    name: wildfly_messaging_$2
+    labels:
+      name: $1
```
Comment (brian-brazil) [inline example_configs/wildfly-10.yaml]: Call this queue or topic

File: example_configs/wildfly-10.yaml [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+  - pattern: "^jboss.as<subsystem=datasources, (?:xa-)*data-source=(.+), statistics=(.+)><>(.+):"
+    attrNameSnakeCase: true
+    name: wildfly_datasource_$2_$3
+    labels:
+      name: $1
+
+  - pattern: "^jboss.as<subsystem=transactions><>(number_of_.+):"
```
Comment (brian-brazil) [inline example_configs/wildfly-10.yaml]: I'd not include the number_of_ in the output name. Every metric is the number of something.

File: example_configs/wildfly-10.yaml (hunk 1 above)
Comment (icetrain) [inline example_configs/wildfly-10.yaml]: So remove the whitelist or did I missunderstand you?

**Comment by icetrain** [comment]
Updated with your suggestions, removing the whitelist increases scrape time with ~8x but since it's just an example config i guess that's not a real problem :)

**Comment by brian-brazil** [comment]
Thanks!

**Comment by Starefossen** [comment]
@cfrantsen I am trying to get this working with Wildfly 10 without any luck. I am wondering if you have a complete example you could share? The problem is the connection from `jmx_exporter` to the `jmx` management interface exposed by Wildlfy...

**Comment by cfrantsen** [comment]
I'm using the jmx_exporter as a java agent when running on WildFly. It requires some fiddling with start parameters to get working, below is an excerpt from the start script I'm using:

```
JBOSS_LOGMANAGER_JAR=$(echo $JBOSS_HOME/modules/system/layers/base/org/jboss/logmanager/main/jboss-logmanager-*.jar)
JMX_EXPORTER_JAR=$(echo /opt/prometheus/jmx_exporter/jmx_prometheus_javaagent-*.jar)

if [ -f $JBOSS_LOGMANAGER_JAR -a -f $JMX_EXPORTER_JAR ]; then
    JBOSS_MODULES_SYSTEM_PKGS="$JBOSS_MODULES_SYSTEM_PKGS,org.jboss.logmanager"

    JAVA_OPTS="$JAVA_OPTS -Xbootclasspath/p:$JBOSS_LOGMANAGER_JAR"
    JAVA_OPTS="$JAVA_OPTS -Djava.util.logging.manager=org.jboss.logmanager.LogManager"
    JAVA_OPTS="$JAVA_OPTS -javaagent:$JMX_EXPORTER_JAR=9102:/opt/prometheus/jmx_exporter/wildfly.yaml"
fi
```

**Comment by Starefossen** [comment]
Thanks Christian! Do you have to specify either `hostPort` or `jmxUrl` in the `wildfly.yaml`? I am use the one from your pull request and I only see `jvm_*` metrics. 

Here is a dump of my `http://localhost: 9102/metrics` https://gist.github.com/Starefossen/dbb98398fdd43053cebc29f42d4f14cb

**Comment by cfrantsen** [comment]
No that should not be needed when running it as a java agent. Have you tried a config file without explicit rules, that should give you a huge amount of metrics back.

**Comment by Starefossen** [comment]
I get much more output, but still nothing about Wildfly/Undertow: https://gist.github.com/Starefossen/9a8ff4fe955fe2b4101333a7e5e83da2

**Comment by Starefossen** [comment]
Sorry for bothering; I have made a [minimal Wildfly jmx_exporter example](https://github.com/Starefossen/docker-wildfly-prometheus/blob/master/Dockerfile) using the official Wildfly Docker image and latest version of `jmx_exporter` which also does not show any Wildfly specific metrics.

[Dockerfile](https://github.com/Starefossen/docker-wildfly-prometheus/blob/master/Dockerfile)

```
ADD jboss-logmanager-2.0.6.Final.jar /opt/jboss/wildfly
ADD jmx_prometheus_javaagent-0.9.jar /opt/jboss/wildfly
ADD prometheus-config.yaml /opt/jboss/wildfly

ENV JAVA_OPTS "-Djboss.modules.system.pkgs=org.jboss.byteman,org.jboss.logmanager \
               -Xbootclasspath/p:/opt/jboss/wildfly/jboss-logmanager-2.0.6.Final.jar \
               -javaagent:/opt/jboss/wildfly/jmx_prometheus_javaagent-0.9.jar=31500:/opt/jboss/wildfly/prometheus-config.yaml"
```

You can run this yourself using `docker-compose up`.

**Comment by cfrantsen** [comment]
We are running 0.7.3.298dcdc here right now so maybe it's a problem with recent versions? Perhaps try an older version and if that works create an issue.

**Comment by Starefossen** [comment]
Thanks a million @cfrantsen 💯 I only tried with v0.8 and v0.9, this is now working perfectly in v0.7!

**Comment by cfrantsen** [comment]
Could you perhaps create an issue regarding this as it can be fixed?

**Comment by juliohm1978** [comment]
Anyone got this working for Wildfly 8.2.1? Using @cfrantsen's [example](https://github.com/prometheus/jmx_exporter/pull/87#issuecomment-289409914) it dies with the following message.

```
java.lang.IllegalArgumentException: Failed to load module "org.jboss.logmanager" for handler "FILE"
	at org.jboss.logmanager.config.AbstractPropertyConfiguration.<init>(AbstractPropertyConfiguration.java:67)
	at org.jboss.logmanager.config.HandlerConfigurationImpl.<init>(HandlerConfigurationImpl.java:54)
	at org.jboss.logmanager.config.LogContextConfigurationImpl.addHandlerConfiguration(LogContextConfigurationImpl.java:145)
	at org.jboss.logmanager.PropertyConfigurator.configureHandler(PropertyConfigurator.java:669)
	at org.jboss.logmanager.PropertyConfigurator.configureLogger(PropertyConfigurator.java:579)
	at org.jboss.logmanager.PropertyConfigurator.configure(PropertyConfigurator.java:517)
	at org.jboss.logmanager.PropertyConfigurator.configure(PropertyConfigurator.java:97)
	at org.jboss.logmanager.LogManager.readConfiguration(LogManager.java:300)
	at org.jboss.logmanager.LogManager.readConfiguration(LogManager.java:262)
	at java.util.logging.LogManager$3.run(LogManager.java:399)
	at java.util.logging.LogManager$3.run(LogManager.java:396)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.util.logging.LogManager.readPrimordialConfiguration(LogManager.java:396)
	at java.util.logging.LogManager.access$800(LogManager.java:145)
	at java.util.logging.LogManager$2.run(LogManager.java:345)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.util.logging.LogManager.ensureLogManagerInitialized(LogManager.java:338)
	at java.util.logging.LogManager.getLogManager(LogManager.java:378)
	at java.util.logging.Logger.demandLogger(Logger.java:448)
	at java.util.logging.Logger.getLogger(Logger.java:502)
	at io.prometheus.jmx.shaded.io.prometheus.jmx.JmxCollector.<clinit>(JmxCollector.java:38)
	at io.prometheus.jmx.shaded.io.prometheus.jmx.JavaAgent.premain(JavaAgent.java:47)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at sun.instrument.InstrumentationImpl.loadClassAndStartAgent(InstrumentationImpl.java:386)
	at sun.instrument.InstrumentationImpl.loadClassAndCallPremain(InstrumentationImpl.java:401)
Caused by: java.lang.NoClassDefFoundError: org/jboss/modules/ModuleLoader
	at org.jboss.logmanager.config.AbstractPropertyConfiguration$ModuleFinder.getClassLoader(AbstractPropertyConfiguration.java:466)
	at org.jboss.logmanager.config.AbstractPropertyConfiguration.<init>(AbstractPropertyConfiguration.java:65)
	... 27 more
```

**Comment by juliohm1978** [comment]
Finally got it working. If anyone else is stumbling on similar issues, remember: the following flags MUST ALL be included in JAVA_OPTS.

```
-Djboss.modules.system.pkgs=org.jboss.byteman,org.jboss.logmanager

-Xbootclasspath/p:/opt/jboss/wildfly/modules/system/layers/base/org/jboss/logmanager/main/jboss-logmanager-X.X.X.Final.jar

-javaagent:/path/to/jmx_prometheus_javaagent.jar=NNNN:/path/to/config.yaml
```

Particularly, I was running in circles trying to define JBOSS_MODULES_SYSTEM_PKGS before starting wildfly. By the looks of standalone.conf, it should end up defining the `-Djboss.modules.system.pkgs` flag.

... Except, it isn't for some reason.

`-Djboss.modules.system.pkgs=org.jboss.logmanager` needs to be there. In order for this to work. Addng it to JAVA_OPTS myself resolved the issue.

**Pull request merged**